ssg/rules.py

```diff
@@ -0,0 +1,130 @@
+from __future__ import absolute_import
+from __future__ import print_function
+
+import os
+
+
+from .build_remediations import REMEDIATION_MAP
+
+
+def get_rule_dir_yaml(dir_path):
+    """
+    Returns the path to the yaml metadata for a rule directory,
+    regardless of if it exists.
+    """
+    return os.path.join(dir_path, "rule.yml")
+
+
+def get_rule_dir_id(path):
+    """
+    Returns the ID of a rule directory; correctly handles being passed
+    either the directory path or the yaml metadata path.
+    """
+    dir_path = path
+
+    if path.endswith("rule.yml"):
+        dir_path = os.path.dirname(path)
+
+    return os.path.basename(dir_path)
+
+
+def is_rule_dir(dir_path):
+    """
+    Returns True iff dir_path is a valid rule directory which exists
+
+    To be valid, dir_path must exist and be a directory and the file
+    returned by get_rule_dir_yaml(dir_path) must exist.
+    """
+    rule_yaml = get_rule_dir_yaml(dir_path)
+
+    is_dir = os.path.isdir(dir_path)
+    has_rule_yaml = os.path.exists(rule_yaml)
+
+    return is_dir and has_rule_yaml
+
+
+def get_rule_dir_ovals(dir_path, product=None):
+    """
+    Gets a list of OVALs contained in a rule directory. If product is
+    None, returns all OVALs. If product is not None, returns applicable
+    OVALs in order of priority:
+
+        {{{ product }}}.xml -> shared.xml
+
+    Only returns OVALs which exist.
+    """
+
+    if not is_rule_dir(dir_path):
+        return []
+
+    oval_dir = os.path.join(dir_path, "oval")
+    has_oval_dir = os.path.isdir(oval_dir)
+    if not has_oval_dir:
+        return []
+
+    if not product:
+        results = []
+        for oval_file in os.listdir(oval_dir):
+            _, ext = os.path.splitext(oval_file)
+            if ext == ".xml":
+                results.append(os.path.join(oval_dir, oval_file))
+
+        return results
+
+    results = []
+    for oval_file in ["%s.xml" % product, "shared.xml"]:
+        oval_path = os.path.join(oval_dir, oval_file)
+        if os.path.isfile(oval_path):
+            results.append(oval_path)
+
+    return results
+
+
+def get_rule_dir_remediations(dir_path, remediation_type, product=None):
```

Comment:
Ditto, the best thing would be to somehow reuse the code pattern (but such requirement is out of scope of this PR).

Reply:
Yep, this is updated. :)